applications/RANSApplication/custom_strategies/generic_residual_based_bossak_scalar_transport_scheme.h

```diff
@@ -0,0 +1,361 @@
+        double C0;
+        double C2;
+        double C3;
+    };
+
+    std::vector<LocalSystemVectorType> mSecondDerivativeValuesVectorOld;
+    std::vector<LocalSystemVectorType> mSecondDerivativeValuesVector;
+    std::vector<LocalSystemMatrixType> mMassMatrix;
+    std::vector<LocalSystemMatrixType> mDampingMatrix;
+
+    const double mAlphaBossak;
+
+    const Variable<double>& mrScalarVariable;
+    const Variable<double>& mrScalarRateVariable;
+    const Variable<double>& mrRelaxedScalarRateVariable;
+
+    BossakConstants mBossak;
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    static void CalculateBossakConstants(BossakConstants& rBossakConstants,
+                                         const double Alpha,
+                                         const double DeltaTime)
+    {
+        TimeDiscretization::Bossak bossak(Alpha, 0.25, 0.5);
+        rBossakConstants.Alpha = bossak.GetAlphaM();
+        rBossakConstants.Gamma = bossak.GetGamma();
+
+        rBossakConstants.C0 =
+            (1.0 - rBossakConstants.Alpha) / (rBossakConstants.Gamma * DeltaTime);
+        rBossakConstants.C2 = 1.0 / (rBossakConstants.Gamma * DeltaTime);
+        rBossakConstants.C3 = (1.0 - rBossakConstants.Gamma) / rBossakConstants.Gamma;
+    }
+
+    template <class TItemType>
+    void AddMassMatrixToRHS(TItemType& rItem, LocalSystemVectorType& rRHS_Contribution, const int ThreadId)
+    {
+        rItem.GetSecondDerivativesVector(mSecondDerivativeValuesVector[ThreadId], 0);
+        (mSecondDerivativeValuesVector[ThreadId]) *= (1.00 - mBossak.Alpha);
+        rItem.GetSecondDerivativesVector(mSecondDerivativeValuesVectorOld[ThreadId], 1);
+        noalias(mSecondDerivativeValuesVector[ThreadId]) +=
+            mBossak.Alpha * mSecondDerivativeValuesVectorOld[ThreadId];
+
+        noalias(rRHS_Contribution) -=
+            prod(mMassMatrix[ThreadId], mSecondDerivativeValuesVector[ThreadId]);
+    }
+
+    template <class TItemType>
+    void CalculateDynamicSystem(TItemType& rItem,
+                                LocalSystemMatrixType& rLHS_Contribution,
+                                LocalSystemVectorType& rRHS_Contribution,
+                                typename TItemType::EquationIdVectorType& rEquationIdVector,
+                                const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        BaseType::CalculateSystemContributions(rItem, rLHS_Contribution, rRHS_Contribution,
+                                               rEquationIdVector, rCurrentProcessInfo);
+
+        const int k = OpenMPUtils::ThisThread();
+
+        rItem.CalculateMassMatrix(mMassMatrix[k], rCurrentProcessInfo);
+        // adding mass contribution to the dynamic stiffness
+        if (mMassMatrix[k].size1() != 0) // if M matrix declared
+        {
+            AddMassMatrixToRHS<TItemType>(rItem, rRHS_Contribution, k);
+
+            noalias(rLHS_Contribution) += mBossak.C0 * mMassMatrix[k];
+        }
```

Comment:
Shouldn't we do something if the if statement is not satisfied?

Reply:
We dont have to, this is merely to support Conditions also using the same method. Cconditions returns zero sized mass matrix by default.